docs/source/daml/reference/packages.rst

```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
+.. code-block:: daml
 
-.. code-block:: yaml
+  import Foo
+
+Sometimes you will have multiple packages with the same exported module name. In that case, you must also specify which package the module comes from, as follows:
+
+.. code-block:: daml
+
+  import "foo" Foo
+
+By default, all modules of ``foo`` are made available when importing ``foo`` as a dependency. To control which modules of ``foo`` get exported, you may add an ``exposed-modules`` field in the ``daml.yaml`` file for ``foo``:
 
-  sdk-version: 0.0.0
-  name: foo
-  source: daml
-  version: 1.0.0
   exposed-modules:
-    - Some.Module
-    - Some.Other.Module
+  - Foo
+
+**Important Limitation:** For DAML archive imports via ``dependencies`` to work, the archive should be compiled with the same DAML SDK version. Otherwise, compilation will fail. If matching the SDK version is not possible, see the next section.
+
+Importing a DAML archive via data-dependencies
+**********************************************
+
+A secondary method for importing a DAML archive, which can be used when the DAML SDK versions do not match, is to import a DAML archive via the ``data-dependencies`` field in ``daml.yaml``:
+
+.. code-block:: yaml
+
   dependencies:
-    - daml-prim
-    - daml-stdlib
-    - /home/johndoe/bar/.daml/dist/bar.dar
+  - daml-prim
+  - daml-stdlib
   data-dependencies:
-    - /home/jondoe/bar-0.0.0/.daml/dist/bar-0.0.0.dar
+  - ../foo/foo.dar
 
-Modules from data dependencies can be imported as usual, but need to be qualified by the
-(generated) instances package:
+You can also import a ``.dalf`` file via data-dependencies.
 
-.. code-block:: daml
+When importing packages this way, the DAML compiler will attempt to reconstruct the DAML interface from the compiled DAML-LF binaries included in the DAML archive.
+
+Compared to dependencies, there are certain disadvantages to this method.
+
+The first disadvantage is that the reconstruction process used for data-dependencies is slower than the direct import process used for dependencies, so it will negatively affect the speed of compilation.
+
+The second disadvantage, which has far-reaching consequences, is that not everything can be perfectly reconstructed via data-dependencies. In particular:
+
+#. Export lists cannot be reconstructed, so imports via data-dependencies can access definitions that were originally hidden. This means it is up to the importing module to respect the data encapsulation of the original module. On the positive side, the encapsulation can also be ignored on purpose, to facilitate upgrades of DAML models to newer SDK versions.
+
+#. Certain advanced type system features also cannot be reconstructed, as they are erased in the process of compiling DAML LF binaries. This includes the ``DataKinds``, ``DeriveGeneric``, and ``FunctionalDependencies`` extensions from GHC. This may result in some definitions being unavailable when importing a module that uses these advanced features.
```

Comment:
I think `DataKinds` don’t work in general anyway? We special case a very small number of kinds, primarily generics, in https://github.com/digital-asset/daml/blob/master/compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs#L1606 but I’m not sure it works at all outside of that.

For FunctionalDepencies, I believe things work as long as we can reexport the typeclass?